### DISCUSSION TITLE
Post Quantum support?

### DISCUSSION THREAD

**Discussion by kuehne-trustable-de** [xipki/xipki#282] [Ideas]
Is there already support for PQC algorithms? Hybrid certificates would be most interesting!

**Comment by xipki** [comment]
PQC support is not in plan. There are still no standards for PQC in certificates.

**Reply by ralienpp** [reply]
NIST standardized the new PQ algorithms in August, now we have ML-DSA and SLH-DSA, they are also supported by BouncyCastle.

- https://csrc.nist.gov/pubs/fips/204/final
- https://csrc.nist.gov/pubs/fips/205/final
- https://csrc.nist.gov/pubs/fips/203/final (ML-KEM, this one is for key encapsulation, not for digital signatures)


Can you consider adding these to the roadmap?

**Comment by flako-dd** [comment]
The important part is the actual standardisation of how PQC-Algorithms get integrated into the existing X.509 ecosystem. Multiple RFCs and Standards are being written, some possibly relevant RFCs are:

[Multiple Public-Key Algorithm X.509 Certificates](https://datatracker.ietf.org/doc/draft-truskovsky-lamps-pq-hybrid-x509/02/)
[Composite ML-DSA for use in Internet PKI](https://datatracker.ietf.org/doc/draft-ounsworth-pq-composite-sigs/)
[Composite ML-KEM for use in X.509 Public Key Infrastructure and CMS](https://datatracker.ietf.org/doc/draft-ietf-lamps-pq-composite-kem/05/)
[Hybrid Public Key Encryption RFC 9180](https://datatracker.ietf.org/doc/rfc9180/)

**Comment by kuehne-trustable-de** [comment]
Hi flako,


the worst algorithm from the CA implementation perspective (the stateful 
ones) are not required by NIST. Hybrid and composite algorithms are 
already supported by bouncy castle.
Ignoring the state  of broader adaption for now, support for PQC in 
xipki shouldn't be a big problem.

Greetings,

Andreas
> The important part is the actual standardisation of how PQC-Algorithms get integrated into the existing X.509 ecosystem. Multiple RFCs and Standards are being written, some possibly relevant RFCs are:
>
> [Multiple Public-Key Algorithm X.509 Certificates](https://datatracker.ietf.org/doc/draft-truskovsky-lamps-pq-hybrid-x509/02/)
> [Composite ML-DSA for use in Internet PKI](https://datatracker.ietf.org/doc/draft-ounsworth-pq-composite-sigs/)
> [Composite ML-KEM for use in X.509 Public Key Infrastructure and CMS](https://datatracker.ietf.org/doc/draft-ietf-lamps-pq-composite-kem/05/)
> [Hybrid Public Key Encryption RFC 9180](https://datatracker.ietf.org/doc/rfc9180/)
>

-- 
Andreas Kühne

Chair of OASIS DSS-X
  
phone: +49 177 293 24 97
mailto: ***@***.***


Trustable solutions UG (haftungsbeschränkt), Gartenheimstr. 39C, 30659 Hannover, Amtsgericht Hannover HRB 219338

**Comment by xipki** [comment]
Hi Andreas and Flako,

I am still waiting for both drafts to be finalized:

https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/
https://datatracker.ietf.org/doc/draft-ietf-lamps-kyber-certificates/

Lijun

**Reply by ralienpp** [reply]
I understand the sentiment - waiting a bit until things settle down avoids unnecessary effort.

On the other hand - early betas with experimental support enable others to begin prototyping and get a rough idea of how it would work. At the moment EJBCA is the only open source CA that can be used for prototyping PQ PKIs (that I am aware of), but I would personally be interested in more diversity here, so as to not keep all my eggs in one basket.

There is an ongoing effort of interop testing at the IETF hackathon, where different implementations test each other's certificates: https://github.com/IETF-Hackathon/pqc-certificates. In the case of ML-DSA
https://ietf-hackathon.github.io/pqc-certificates/pqc_hackathon_results_certs_r4.html everyone's parser is able to check everyone else's certificate, so you could say that a stable equilibrium was reached. Things can still change, but I think the future will be a minor variation of the current state (i.e., no dramatic changes will occur).

**Comment by kuehne-trustable-de** [comment]
Hi Lijun,
> I am still waiting for both drafts to be finalized:
>
> https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/
> https://datatracker.ietf.org/doc/draft-ietf-lamps-kyber-certificates/

yes, that's a valid point. I suffered myself from inconsistent and 
changing OIDs.

On the other hand: Do you really have to deal with the OIDs? I would use 
BCs definitions.


Greetings,


Andreas

**Reply by xipki** [reply]
Support of PQC and composite PQC have been added in the master branch.
